### PR TITLE
Fix HttpPath.resolve(String) bugs

### PR DESCRIPTION
**Pull request by lbergelson**
* fix a number of bugs involving relative HttpPaths and the resolve methods when using strings

### CODE REVIEW

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java [hunk 1]
```diff
@@ -634,7 +637,7 @@ public void testNormalizeSeveralSlashes(final String withSlashes, final String w
     public Object[][] toResolve(){
         return new Object[][]{
             {getHttpPath("http://hello.com"), null, getHttpPath("http://hello.com")},
-            {getHttpPath("http://hello.com"), getHttpPath("http://goodbye.com"), getHttpPath("http://goodbye.com")},
+        //    {getHttpPath("http://hello.com"), getHttpPath("http://goodbye.com"), getHttpPath("http://goodbye.com")},  unsupproted
```
Comment (droazen) [inline src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java]: Add comment explaining what is unsupported

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java [hunk 2]
```diff
@@ -694,9 +697,22 @@ public void testResolveString(HttpPath path, String resolveString, HttpPath expe
...
 
-    @Test(expectedExceptions = HttpPath.CantDealWithThisException.class)
-    public void testWeGiveUpOnHorribleInput(){
-        getHttpPath("http://hello.com/hi.there").resolve("http://goodbye.com");
+    @DataProvider
+    public Object[][] absoluteStringsToResolveAgainst() {
+            return new Object[][]{
+                    {"http://example.com"},
+                    {"http://example.com/file.txt"},
+                    {"https://example.com"},
```
Comment (droazen) [inline src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java:705]: Is there any value in adding `"https://example.com/"` as an additional case?

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java [hunk 3]
```diff
@@ -694,9 +697,22 @@ public void testResolveString(HttpPath path, String resolveString, HttpPath expe
...
+                    {"https://example.com/file.txt"},
+                    {"file:///local"},
+                    {"c://local"},
+                    {"file:///local/file"}
+            };
+    }
+
+    @Test(dataProvider = "absoluteStringsToResolveAgainst", expectedExceptions = UnsupportedOperationException.class)
+    public void testResolveAgainstAbsoluteFailsStrings(String other){
+        getHttpPath("http://www.example.com/file.txt").resolve(other);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java:716]: Do you also want to try against a URI with just a domain? Eg., "http://www.example.com/"?

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java [hunk 4]
```diff
@@ -716,10 +732,16 @@ public Object[][] getSiblingTests() {
     }
```
Comment (droazen) [inline src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java:738]: Why does the anchor go away in the expected results here?

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java [hunk 5]
```diff
@@ -729,8 +751,42 @@ public void testNullSiblingIsNPE(){
...
+        Assert.assertEquals(subpath.toString(), expected);
+    }
+
+    // From htsjdk reported failure
+    @Test
+    public void testResolveFasta() {
+        Path fastaFile = getHttpPath("https://example.com/fastas/fasta.gz");
+        final Path fileNamePath = fastaFile.getFileName();
+        final String indexName = fileNamePath + ".fai";
+        fastaFile.resolveSibling(indexName);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java]: Check the return value of `resolveSibling()` here

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java [hunk 6]
```diff
@@ -729,8 +751,42 @@ public void testNullSiblingIsNPE(){
...
+                {absolutePathWithQueryAndReference.subpath(3, 4), "fasta.gz"},
+        };
+    }
+
+    @Test(dataProvider = "relativePaths")
+    public void testRelativePathsToString(Path subpath, String expected){
+        Assert.assertEquals(subpath.toString(), expected);
+    }
+
+    // From htsjdk reported failure
```
Comment (droazen) [inline src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java]: Link to github issue

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java [hunk 7]
```diff
@@ -729,8 +751,42 @@ public void testNullSiblingIsNPE(){
...
+    @Test
+    public void testResolveFasta() {
+        Path fastaFile = getHttpPath("https://example.com/fastas/fasta.gz");
+        final Path fileNamePath = fastaFile.getFileName();
+        final String indexName = fileNamePath + ".fai";
+        fastaFile.resolveSibling(indexName);
+    }
+
+    @Test
+    public void testWithHtsjdkFindIndex(){
```
Comment (droazen) [inline src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java:795]: Explain what bug this is testing

File: src/main/java/org/broadinstitute/http/nio/HttpPath.java [hunk 8]
```diff
@@ -352,80 +351,73 @@ public Path normalize() {
...
-        } else if(other.isAbsolute()){
-            if(other instanceof HttpPath absolutePath){
-                return absolutePath;
-            } else {
-                throw new IllegalArgumentException("Cannot a resolve an absolute path which isn't an HttpPath against an" +
-                        " HttpPath."
-                        + "\nThis path is: " + this.toUriString()
+        }  else if(other.isAbsolute()){
+            //Note: This violates the general contract of the method but shouldn't be important practically.
+                throw new UnsupportedOperationException("Cannot a resolve an absolute path against an http(s)."
```
Comment (droazen) [inline src/main/java/org/broadinstitute/http/nio/HttpPath.java]: `Cannot a` -> `Cannot`
`http(s)` -> `http(s) path`

File: src/main/java/org/broadinstitute/http/nio/HttpPath.java [hunk 9]
```diff
@@ -352,80 +351,73 @@ public Path normalize() {
...
+                        + "\nThis path is: " + this
                         + "\nThe problematic path is an instance of " + other.getClass().getName()
                         + "\nOther path: " + other);
-            }
         } else {
-            final URI otherUri;
             try {
-                otherUri = new URI(other.toString());
+                final URI otherUri = new URI(other.toString());
```
Comment (droazen) [inline src/main/java/org/broadinstitute/http/nio/HttpPath.java]: Document why we're using `toString()` here

File: src/main/java/org/broadinstitute/http/nio/HttpPath.java [hunk 10]
```diff
@@ -620,7 +619,6 @@ private void initOffsets() {
      * Gets the path as a normalized (without multiple slashes) array of bytes.
      *
      * @param path          path to convert into byte[]
-     * @param checkRelative if {@code true}, check if the path is absolute.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/http/nio/HttpPath.java]: Param is still there?

File: src/main/java/org/broadinstitute/http/nio/HttpPath.java [hunk 11]
```diff
@@ -352,80 +351,73 @@ public Path normalize() {
...
     @Override
     public HttpPath resolve(final String other) {
-        return resolve(other == null ? null: Paths.get(other));
+        return resolve(fromRelativeString(other));
     }
 
     @Override
     public Path resolveSibling(final String other){
-        return resolveSibling(other == null ? null : Paths.get(other));
+        return resolveSibling(fromRelativeString(other));
```
Comment (droazen) [inline src/main/java/org/broadinstitute/http/nio/HttpPath.java:394]: Is it worth adding a comment mentioning that we can't use `Paths.get()` here and must go through the `fromRelativeString()` codepath?

**Review by droazen — APPROVED** [review]
@lbergelson Minor comments only -- merge/release after addressing.

File: src/main/java/org/broadinstitute/http/nio/HttpPath.java (hunk 8 above)
Comment (lbergelson) [reply src/main/java/org/broadinstitute/http/nio/HttpPath.java]: done

File: src/main/java/org/broadinstitute/http/nio/HttpPath.java (hunk 9 above)
Comment (lbergelson) [reply src/main/java/org/broadinstitute/http/nio/HttpPath.java]: added

File: src/main/java/org/broadinstitute/http/nio/HttpPath.java (hunk 11 above)
Comment (lbergelson) [reply src/main/java/org/broadinstitute/http/nio/HttpPath.java:394]: commentted

File: src/main/java/org/broadinstitute/http/nio/HttpPath.java (hunk 10 above)
Comment (lbergelson) [reply src/main/java/org/broadinstitute/http/nio/HttpPath.java]: resurrected

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java (hunk 1 above)
Comment (lbergelson) [reply src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java]: explained

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java (hunk 2 above)
Comment (lbergelson) [reply src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java:705]: added, probably not important

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java (hunk 3 above)
Comment (lbergelson) [reply src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java:716]: I don't think it's important but it's trivially testable so I added it

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java (hunk 4 above)
Comment (lbergelson) [reply src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java:738]: To be continued....

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java (hunk 5 above)
Comment (lbergelson) [reply src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java]: done

File: src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java (hunk 7 above)
Comment (lbergelson) [reply src/test/java/org/broadinstitute/http/nio/HttpPathUnitTest.java:795]: done

**Review by lbergelson — COMMENTED** [review]
responded to review

**Pull request merged**